ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java

```diff
@@ -116,17 +116,20 @@ public int end (boolean report_heuristics)
 	 */
 	public int cancel ()
 	{
-		if (TxStats.enabled())
-			TxStats.getInstance().incrementApplicationRollbacks();
-
 		if (parent() != null)
 			parent().removeChildAction(this);
 
 		// beforeCompletion();
 
-		int outcome = super.Abort();
+		int outcome;
+		synchronized (_cancelLock) { // https://issues.jboss.org/browse/JBTM-2643
```

Comment:
I don't like having yet another lock in here. Why not put the conditional stat increment in BasicAction::Abort which is already synchronized.